source/targetfiltersource_test.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/context"
+
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+type mockTargetNetFilter struct {
+	targets map[string]bool
+}
+
+func NewMockTargetNetFilter(targets []string) endpoint.TargetFilterInterface {
+	targetMap := make(map[string]bool)
+	for _, target := range targets {
+		targetMap[target] = true
+	}
+	return &mockTargetNetFilter{targets: targetMap}
+}
+
+func (m *mockTargetNetFilter) Match(target string) bool {
+	return m.targets[target]
+}
+
+func (m *mockTargetNetFilter) IsConfigured() bool {
+	return true
+}
+
+// echoSource is a Source that returns the endpoints passed in on creation.
+type echoSource struct {
+	endpoints []*endpoint.Endpoint
+}
+
+func (e *echoSource) AddEventHandler(ctx context.Context, handler func()) {
+}
+
+// Endpoints returns all of the endpoints passed in on creation
+func (e *echoSource) Endpoints(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	return e.endpoints, nil
+}
+
+// NewEchoSource creates a new echoSource.
+func NewEchoSource(endpoints []*endpoint.Endpoint) Source {
+	return &echoSource{endpoints: endpoints}
+}
+
+func TestEchoSourceReturnGivenSources(t *testing.T) {
+	startEndpoints := []*endpoint.Endpoint{{
+		DNSName:    "foo.bar.com",
+		RecordType: "A",
+		Targets:    endpoint.Targets{"1.2.3.4"},
+		RecordTTL:  endpoint.TTL(300),
+		Labels:     endpoint.Labels{},
+	}}
+	e := NewEchoSource(startEndpoints)
+
+	endpoints, err := e.Endpoints(context.Background())
+	if err != nil {
+		t.Errorf("Expected no error but got %s", err.Error())
+	}
+
+	for i, endpoint := range endpoints {
+		if endpoint != startEndpoints[i] {
+			t.Errorf("Expected %s but got %s", startEndpoints[i], endpoint)
+		}
+	}
+}
+
+func TestTargetFilterSource(t *testing.T) {
+	t.Parallel()
+
+	t.Run("Interface", TestTargetFilterSourceImplementsSource)
+	t.Run("Endpoints", TestTargetFilterSourceEndpoints)
+}
+
+// TestTargetFilterSourceImplementsSource tests that targetFilterSource is a valid Source.
+func TestTargetFilterSourceImplementsSource(t *testing.T) {
+	assert.Implements(t, (*Source)(nil), new(targetFilterSource))
```

Comment:
FYI, a simpler way of writing this is `var _ Source = &targetFilterSource{}`

Reply:
The actual assert? i.e. replace `assert.Implements(t, (*Source)(nil), new(targetFilterSource))` with `var _ Source = &targetFilterSource{}`?

I believe this was just straight copied from `service_test.go` and changed, I do however assume that assert would have some kind of helpful fail message?

Reply:
No, in place of the assert. If `targetFilterSource` doesn't implement `Source` then there would be a compilation failure with a helpful message.